src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,173 @@
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            long streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _streamCancellationToken = cancellationToken;
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize / 2));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // Ideally we'd have IAsyncEnumerable<ReadOnlySequence<byte>> here so we can pass through the
+        // data without having to copy it into a temporary buffer in BlazorPackHubProtocolWorker. But trying
+        // this gives strange errors; sometimes the "chunk" variable below has a negative length, even though
+        // the logic never returns a corrupted item as far as I can tell.
+        private async Task<bool> ReceiveData(byte[] chunk, string error)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(error))
+                {
+                    throw new InvalidOperationException($"An error occurred while reading the remote stream: {error}");
+                }
+
+                if (chunk.Length == 0)
+                {
+                    throw new InvalidOperationException($"The incoming data chunk cannot be empty.");
+                }
+
+                _bytesRead += chunk.Length;
+
+                if (_bytesRead > _totalLength)
+                {
+                    throw new InvalidOperationException($"The incoming data stream declared a length {_totalLength}, but {_bytesRead} bytes were read.");
+                }
```

Comment:
We need to think about this exceptions in terms of security. These are influenced by "user" (the browser) input. We need to understand which of these can happen naturally and which ones would only happen if a client is not behaving correctly.

Exceptions caused by a client not behaving correctly should result on the circuit being immediately terminated and a debug entry written to the log as opposed to a warning or an error. Otherwise an external actor can fill our logs with warnings/errors.

I'm not suggesting you do a change at the moment, however its something we need to revisit before completing this work.

When can you receive an `error`? Is it something recoverable?

Reply:
> When can you receive an error? Is it something recoverable?

Essentially [when we run into errors](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-7c47c684c78e3f5b5eeecb28b249b03afcefcd90a925cc20b69ec2cdec261752R177
) sending chunks (maximum SignalR message size is too small, issue with invoke/send). It is NOT recoverable.

Reply:
> Essentially [when we run into errors](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-7c47c684c78e3f5b5eeecb28b249b03afcefcd90a925cc20b69ec2cdec261752R177) sending chunks (maximum SignalR message size is too small

We should never run into issues with this (we can compute the correct value because we know the limit ahead of time)

Reply:
> We should never run into issues with this (we can compute the correct value because we know the limit ahead of time as in SignalROptions.MaxMessageSize or whatever is called, that can be flowed to the client on the OpenRead call)

So we do flow this value through already in the OpenRead call (see discussion [here](https://github.com/dotnet/aspnetcore/pull/33491#discussion_r653780233)). However there is a potential (highly unlikely) case that the max SignalR message size has been set to 100 bytes, which would cause the client to throw an error as we allocate at least 512 bytes to overhead.

Reply:
> However there is a potential (highly unlikely) case that the max SignalR message size has been set to 100 bytes

If we are concerned about this, can we detect this on the server and throw an exception before we even start sending data? It's not something we care about in general.

Reply:
> If we are concerned about this, can we detect this on the server and throw an exception before we even start sending data? It's not something we care about in general.

Updated accordingly. 

I'd still like to keep the client provided `error` to the `ReceiveJSDataChunk` in case of exceptionary circumstances (ex. something goes wrong with SignalR causing the client to error out). By having the client send the error, the server can proceed to close the pipe and dispose the stream. Let me know if you have concerns with this approach.

Reply:
Regarding the user provided `error` message we catch the [exception](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-58549a26a6d664614a7f6cabf5444c9732521dfb9150bf198614f1da901b4577R95) (without directly logging) and [pass it through the `_pipe.Writer`](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-58549a26a6d664614a7f6cabf5444c9732521dfb9150bf198614f1da901b4577R198). If the developer has not caught this exception when they read from the `RemoteJSDataStream`, then it'll be a fatal exception which may be logged. 

For other exceptions (out of order chunk, 0 length chunk, providing more data than expected, [we treat it as a fatal exception and rethrow](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-58549a26a6d664614a7f6cabf5444c9732521dfb9150bf198614f1da901b4577R130-R135). Consequently, this is [logged](https://github.com/dotnet/aspnetcore/pull/33491/files#diff-c08c874d10203eb22ffb6d6417cb0fcc90d4984f9486ae7c585024b458e06b5fR434-R437) (at a debug level).